src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationContext.cs

```diff
@@ -1335,53 +1343,69 @@ private static DkmClrCompilationResultFlags GetLocalResultFlags(LocalSymbol loca
         {
             // Display class instance. The display class fields are variables.
             int n = 0;
+
             foreach (var member in instance.Type.GetMembers())
             {
                 if (member.Kind != SymbolKind.Field)
                 {
                     continue;
                 }
+
                 var field = (FieldSymbol)member;
                 var fieldType = field.Type;
-                var fieldKind = GeneratedNames.GetKind(field.Name);
-                if (fieldKind == GeneratedNameKind.DisplayClassLocalOrField ||
-                    fieldKind == GeneratedNameKind.TransparentIdentifier ||
-                    IsTransparentIdentifierFieldInAnonymousType(field) ||
-                    (fieldKind == GeneratedNameKind.ThisProxyField && GeneratedNames.GetKind(fieldType.Name) == GeneratedNameKind.LambdaDisplayClass)) // Async lambda case.
+                var fieldName = field.Name;
+                GeneratedNameKind fieldKind;
+                int openBracketOffset;
+                int closeBracketOffset;
+                GeneratedNames.TryParseGeneratedName(fieldName, out fieldKind, out openBracketOffset, out closeBracketOffset);
+
+                switch (fieldKind)
                 {
-                    Debug.Assert(!field.IsStatic);
-                    // A local that is itself a display class instance.
-                    if (displayClassTypes.Add((NamedTypeSymbol)fieldType))
-                    {
-                        var other = instance.FromField(field);
-                        displayClassInstances.Add(other);
-                        n++;
-                    }
+                    case GeneratedNameKind.DisplayClassLocalOrField:
+                    case GeneratedNameKind.TransparentIdentifier:
+                        break;
+                    case GeneratedNameKind.AnonymousTypeField:
+                        var part = fieldName.Substring(openBracketOffset + 1, closeBracketOffset - openBracketOffset - 1);
```

Comment:
Added a helper to this class rather than to `GeneratedNames` since the helper is simple and there is only one call site outside the EE, and it may be more confusing to try to document a general helper than to implement twice.